src/macaron/slsa_analyzer/provenance/intoto/v1/__init__.py

```diff
@@ -12,3 +44,211 @@ class InTotoV1Statement(TypedDict):
+        When the payload does not follow the expecting schema.
+    """
+    if not isinstance(subject, dict):
+        raise ValidateInTotoPayloadError(
+            "A subject in the in-toto statement is invalid: expecting an object.",
+        )
+
+    # At least one of 'uri', 'digest', and 'content' must be valid and present.
+    uri_state = _validate_property(subject, "uri", True, lambda x: isinstance(x, str), False)
+    content_state = _validate_property(subject, "content", True, lambda x: isinstance(x, str), False)
+    digest_state = _validate_property(subject, "digest", True, is_valid_digest_set, False)
+    if _ValidityState.VALID not in (uri_state, content_state, digest_state):
+        raise ValidateInTotoPayloadError(
+            f"One of 'uri', 'digest', or 'content' must be present and valid within 'subject'. URI: {uri_state.name}. "
+            f"Content: {content_state.name}. DIGEST: {digest_state.name}."
+        )
+
+    _validate_property(subject, "name", False, lambda x: isinstance(x, str))
+    _validate_property(subject, "downloadLocation", False, lambda x: isinstance(x, str))
+    _validate_property(subject, "mediaType", False, lambda x: isinstance(x, str))
+    _validate_property(subject, "annotations", False, is_valid_annotation_map)
+
+    return True
+
+
+def is_valid_digest_set(digest: JsonType) -> bool:
+    """Validate the digest set.
+
+    Specification for the digest set: https://github.com/in-toto/attestation/blob/main/spec/v1/digest_set.md.
+
+    Parameters
+    ----------
+    digest : JsonType
+        The digest set.
+
+    Returns
+    -------
+    bool:
+        ``True`` if the digest is valid according to the spec.
+    """
+    if not isinstance(digest, dict):
+        return False
+    for key in digest:
+        if key not in VALID_ALGORITHMS:
+            return False
+        if not isinstance(digest[key], str):
+            return False
+    return True
+
+
+def is_valid_annotation_map(annotation_map: JsonType) -> bool:
+    """Validate the annotation map which is a dictionary with string keys and JsonType values.
+
+    Parameters
+    ----------
+    annotation_map : JsonType
+        The annotation map dictionary.
+
+    Returns
+    -------
+    bool:
+        ``True`` if the annotation map is valid according to the spec.
+    """
+    if not isinstance(annotation_map, dict):
+        return False
+    return True
+
+
+def _validate_property(
+    object_: JsonType,
+    key: str,
+    required: bool,
+    validator: Callable[[JsonType], bool],
+    report_errors: bool = True,
+) -> _ValidityState:
+    """Validate the existence and type of target within the passed Json object."""
+    if not isinstance(object_, dict):
+        # The passed object is checked before this function is called, but mypy requires it be explicitly checked here
+        #  too. Also known as: this branch should never be taken.
+        return _ValidityState.INVALID
+
+    value = object_.get(key)
```

Comment:
Since you have already checked if `subject` is a `dict`, `object_` here can be type-annotated as `dict`. That way you don't need the `if` statement to narrow the type again to make `mypy` happy.

```suggestion
def _validate_property(
    object_: dict[str, JsonType],
    key: str,
    required: bool,
    validator: Callable[[JsonType], bool],
    report_errors: bool = True,
) -> _ValidityState:
    """Validate the existence and type of target within the passed Json object."""
   value = object_.get(key)
```